### ISSUE TITLE
Batchnorm parameters not gets trained when training

### ISSUE THREAD

**Issue by MrWanter** [ppwwyyxx/tensorpack#696]
Hi I used your implementation of wrapper of Conv2D function as :
```
        with argscope([Conv2D, MaxPooling, AvgPooling, GlobalAvgPooling, BatchNorm], data_format='NCHW'), \
                argscope(Conv2D, use_bias=False):
            with TowerContext('', is_training=True):
```
then a lot of `Conv2D` func like:
`Conv2D('conv1', l, 16, 3, stride=1, nl=BNReLU)`

but the saved model got `0` for` bn/mean` and 1 for `bn/variance`, seems not got trained, could you pls suggest me where I am wrong?

**Comment by MrWanter** [comment]
turns out updated op needs to be added with training

**Comment by ppwwyyxx** [comment]
Alternatively you can use the `internal_update` option.

**Comment by MrWanter** [comment]
what if I want to let some layer's batchnorm act as testing state(use trained moving averages), while other layers as training state(use batch statistics)?

**Comment by MrWanter** [comment]
can I do:
```
with TowerContext('', is_training=True):
        Conv2D(...)
with TowerContext('', is_training=False):
        Conv2D(...)
```

**Comment by MrWanter** [comment]
but I only use the `Conv2D` which used` nl=BNRelu or nl= BN `
for the 
```
with TowerContext('', is_training=False):
        Conv2D(...)
```
part, I want to load pretrained batchnorm params(moving average)
for the 
```
with TowerContext('', is_training=True):
        Conv2D(...)
```
part, I want to train it as usual(use the batch statistics and keep moving average)
would the my way of 
```
with TowerContext('', is_training=True):
        Conv2D(...)
with TowerContext('', is_training=False):
        Conv2D(...)
```
work? Thank you

**Comment by ppwwyyxx** [comment]
It would not work and you're supposed to see an error.

You can use a different argscope around each Conv2D. But it's better to not use BNReLU or reimplement BNReLU if you have complicated use cases.

**Comment by MrWanter** [comment]
but I didn't see an error, will this work:
```
with argscope([Conv2D, BatchNorm], is_training=False):
      Conv2D(...,nl=BNRelu)
with argscope([Conv2D, BatchNorm], is_training=True):
      Conv2D(...,nl=BNRelu)
```

**Comment by ppwwyyxx** [comment]
> with argscope([Conv2D, BatchNorm], is_training=False)

This is not how `argscope` works. `Conv2D` and `BatchNorm` do not accept the `is_training` argument.

**Comment by MrWanter** [comment]
could you pls point out a way to do this? Thanks

**Comment by ppwwyyxx** [comment]
`argscope(BatchNorm, training=False)`

**Comment by MrWanter** [comment]
I see, so this will work?
```
with argscope([Conv2D, BatchNorm], training=False):
      Conv2D(...,nl=BNRelu)
with argscope([Conv2D, BatchNorm], training=True):
      Conv2D(...,nl=BNRelu)
```

**Comment by ppwwyyxx** [comment]
It will not. `Conv2D` does not accept the `training` argument.

```
with argscope(BatchNorm, training=False):
      Conv2D(...,nl=BNRelu)
with argscope(BatchNorm, training=True):
      Conv2D(...,nl=BNRelu)
```

**Comment by MrWanter** [comment]
but seems Conv2D also dont accept training argument
```
def Conv2D(x, out_channel, kernel_shape,
           padding='SAME', stride=1,
           W_init=None, b_init=None,
           nl=tf.identity, split=1, use_bias=True,
           data_format='NHWC'):
```

**Comment by MrWanter** [comment]
I see, thank you

**Comment by MrWanter** [comment]
but seems batchnorm also dont accept training argument:
```
def BatchNorm(x, use_local_stat=None, decay=0.9, epsilon=1e-5,
              use_scale=True, use_bias=True,
              gamma_init=tf.constant_initializer(1.0), data_format='NHWC',
              internal_update=False):
```

**Comment by ppwwyyxx** [comment]
Then you're using an old version of tensorpack.

**Comment by ppwwyyxx** [comment]
To be honest, there is no benefit in using tensorpack layers if you don't use tensorpack trainers. It only causes more trouble with a set of strange interface.
The `training` option used to be named `use_local_stat`, but recently it was renamed to be consistent with official `tf.layers`.

**Comment by MrWanter** [comment]
so my use of 
```
with TowerContext('', is_training=True):
        Conv2D(...)
with TowerContext('', is_training=False):
        Conv2D(...)
```
in the old version will work? as 
```
        use_local_stat (bool): whether to use mean/var of the current batch or the moving average.
            Defaults to True in training and False in inference.
```

**Comment by ppwwyyxx** [comment]
It will work only if you did not use tensorpack trainers to train the model.

**Comment by MrWanter** [comment]
Yes, I dont use tensorpack trainner
